File: backend/src/main/java/com/team01/backend/global/security/JwtAuthenticationFilter.java
package com.team01.backend.global.security;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

/**
 * [과제] 모든 요청에 대해 JWT 토큰이 유효한지 검사하는 '필터' 클래스입니다.
 * Security 필터 체인에서 사용자 인증을 담당합니다.
 * [수정] 헤더 방식에서 HttpOnly 쿠키 방식으로 토큰 추출 로직이 변경되었습니다.
 */
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    private final JwtTokenProvider jwtTokenProvider;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
            throws ServletException, IOException {

        // 1. [수정] 쿠키 저장소에서 accessToken을 추출합니다.
        String token = resolveTokenFromCookie(request);

        // 2. 토큰이 유효하다면 인증 정보를 SecurityContext에 저장합니다.
        if (token != null && jwtTokenProvider.validateToken(token)) {
            Authentication authentication = jwtTokenProvider.getAuthentication(token);
            SecurityContextHolder.getContext().setAuthentication(authentication);
        }

        // 3. 다음 필터로 요청을 넘깁니다.
        filterChain.doFilter(request, response);
    }

    /**
     * [신규 추가] 쿠키 배열을 뒤져 accessToken 값을 찾아내는 내부 메서드
     */
    private String resolveTokenFromCookie(HttpServletRequest request) {
        Cookie[] cookies = request.getCookies();
        if (cookies != null) {
            for (Cookie cookie : cookies) {
                if ("accessToken".equals(cookie.getName())) {
                    return cookie.getValue();
                }
            }
        }
        return null;
    }
}

File: backend/src/main/java/com/team01/backend/global/security/JwtAuthenticationFilter.kt
package com.team01.backend.global.security

import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpHeaders
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException

/**
 * 모든 API 요청이 들어올 때마다 JWT 토큰의 유효성을 검사하는 필터입니다.
 */
class JwtAuthenticationFilter(private val jwtTokenProvider: JwtTokenProvider) : OncePerRequestFilter() {

    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {

        val token = resolveToken(request)

        if (token != null && jwtTokenProvider.validateToken(token)) {
            // 토큰으로부터 사용자의 인증 정보(Authentication)를 가져옵니다.
            val authentication = jwtTokenProvider.getAuthentication(token)
            // SecurityContext에 인증 정보를 저장하여, 이후 로직에서 인증된 사용자로 인식하게 합니다.
            SecurityContextHolder.getContext().authentication = authentication
        }

        filterChain.doFilter(request, response)
    }

    private fun resolveToken(request: HttpServletRequest): String? {
        return resolveTokenFromHeader(request) ?: resolveTokenFromCookie(request)
    }

    private fun resolveTokenFromHeader(request: HttpServletRequest): String? {
        val bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION)
        return bearerToken
            ?.takeIf { it.startsWith(BEARER_PREFIX) }
            ?.substring(BEARER_PREFIX.length)
            ?.takeIf { it.isNotBlank() }
    }

    private fun resolveTokenFromCookie(request: HttpServletRequest): String? {
        return request.cookies?.find { cookie -> cookie.name == ACCESS_TOKEN_COOKIE }?.value
    }

    companion object {
        private const val BEARER_PREFIX = "Bearer "
        private const val ACCESS_TOKEN_COOKIE = "accessToken"
    }
}